File: problems/0155-min-stack/python/solution.py
class MinStack:
    """Min Stack — https://leetcode.com/problems/min-stack/"""

    def __init__(self):
        self.stack = []
        self.mins = []

    def push(self, val):
        self.stack.append(val)
        self.mins.append(val if not self.mins else min(val, self.mins[-1]))

    def pop(self):
        self.stack.pop()
        self.mins.pop()

    def top(self):
        return self.stack[-1]

    def getMin(self):
        return self.mins[-1]


def minStackOps(ops, args):
    """Test driver: replay LeetCode-style operation/argument lists."""
    stack = MinStack()
    out = []
    for op, arg in zip(ops, args):
        if op == "MinStack":
            out.append(None)
        elif op == "push":
            stack.push(arg[0])
            out.append(None)
        elif op == "pop":
            stack.pop()
            out.append(None)
        elif op == "top":
            out.append(stack.top())
        elif op == "getMin":
            out.append(stack.getMin())
    return out

File: problems/0155-min-stack/typescript/solution.ts
// Min Stack — https://leetcode.com/problems/min-stack/
class MinStack {
  private stack: number[] = [];
  private mins: number[] = [];

  push(val: number): void {
    this.stack.push(val);
    const min = this.mins.length === 0 ? val : Math.min(val, this.mins[this.mins.length - 1]);
    this.mins.push(min);
  }

  pop(): void {
    this.stack.pop();
    this.mins.pop();
  }

  top(): number {
    return this.stack[this.stack.length - 1];
  }

  getMin(): number {
    return this.mins[this.mins.length - 1];
  }
}

// Test driver: replay LeetCode-style operation/argument lists (null for void ops).
export function minStackOps(ops: string[], args: number[][]): (number | null)[] {
  const stack = new MinStack();
  const out: (number | null)[] = [];
  for (let i = 0; i < ops.length; i++) {
    switch (ops[i]) {
      case "MinStack":
        out.push(null);
        break;
      case "push":
        stack.push(args[i][0]);
        out.push(null);
        break;
      case "pop":
        stack.pop();
        out.push(null);
        break;
      case "top":
        out.push(stack.top());
        break;
      case "getMin":
        out.push(stack.getMin());
        break;
    }
  }
  return out;
}
